alphazero/_game.py

```diff
@@ -383,3 +384,339 @@ def n_players():
         Get the number of players in the game.
         """
         return 2
+
+
+class Connect4(Game):
+    def __init__(self, width=7, height=6, board_state=None):
+        """
+        Construct an instance of the game class.
+        """
+        if board_state is None:
+            self._board = self.get_initial_board(height, width)
+        else:
+            tensor = torch.Tensor(board_state)
+            if width * height != tensor.numel():
+                raise ValueError(
+                    f"width and height dimensions {width}, {height} do not "
+                    f"match board state size {tensor.shape}"
+                )
+            self._board = tensor.view(height, width)
+        self._move_count = 0
+        self._players = [0, 1]
+        self.win_list = [False, False]
+
+    def make_move(self, move: int):
+        """
+        Make a move on the board. Should be a valid move and return a valid
+        game board.
+        """
+        self._board = self.board_after_move(
+            self._board, self._players[self._move_count % 2], move
+        )
+        self._move_count += 1
+
+        if not self.is_valid():
+            raise ValueError(f"Board state {self._board} is not valid.")
+
+    @staticmethod
+    def board_after_move(board, player: int, move: int):
+        """
+        Make a move on the board and return it. Does not affect any instances
+        of the class. Should be a valid move and return a valid game board.
+        """
+        if not isinstance(move, int):
+            raise TypeError(f"move is of type {type(move)} but must be an int")
+
+        if player not in (0, 1):
+            raise ValueError(f"invalid player for move {move}")
+
+        if any((board[:, move] == -1)) is False:
+            raise ValueError(f"Column is full")
+        if move > board.shape[0]:
+            raise IndexError(
+                f"invalid move {move} for board shaped {board.shape}"
+            )
+
+        row_index = board.shape[0] - 1
+        while row_index >= 0 and board[row_index, move] != -1:
+            row_index -= 1
```

Comment:
use something like this instead maybe? https://github.com/pytorch/pytorch/issues/9413

Reply:
Sorry not too sure how to implement this.. I'll look into it later, but from my understanding, if we use nonzero, we might get confused with spaces with player = 1. So it might not be usable, since we are trying find the last index with -1 to replace with the current player, and if we grab a space with player = 1, well you get it I guess...

Reply:
gotcha. just wanted to raise the q